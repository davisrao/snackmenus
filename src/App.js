import React, { useState, useEffect } from "react";
import { BrowserRouter } from "react-router-dom";
import "./App.css";
import Home from "./Home";
import SnackOrBoozeApi from "./Api";
import NavBar from "./NavBar";
import { Route, Switch } from "react-router-dom";
import FoodMenu from "./FoodMenu";
import ItemDetails from "./ItemDetails";
import DrinkMenu from "./DrinkMenu";
import MenuItemForm from "./MenuItemForm";
import NotFound from "./NotFound"


/** App component
 * props:
 * - none
 * 
 * state:
 * - isLoading (T/F) 
 * - drinks (array of all drink objects)
 * - snacks (array of all snack objects)
 * 
 * context:
 * - none
 * 
 * App => {NavBar,Routes}
 */

function App() {
  const [isLoading, setIsLoading] = useState(true);
  const [drinks, setDrinks] = useState([]);
  const [snacks, setSnacks] = useState([]);
  console.log("*App", isLoading, drinks, snacks)


  /**Effect running on initial load of data to query snack + drink data */
  useEffect(() => {
    async function getDrinks() {
      const drinks = await SnackOrBoozeApi.getDrinks();
      setDrinks(drinks);
    }
    async function getSnacks() {
      const snacks = await SnackOrBoozeApi.getSnacks();
      setSnacks(snacks);
      setIsLoading(false);
    }
    getDrinks();
    getSnacks();

  }, []);

  /**
   * Function to update data outside of useEffect
   * accepts type, and updates data that was edited
   * if type is snacks, update snacks, if type is drinks, 
   * update drinks
   */
  async function updateData(type) {
    if (type === "drinks") {
      let drinks = await SnackOrBoozeApi.getDrinks();
      setDrinks(drinks);
    } else if (type === "snacks") {
      let snacks = await SnackOrBoozeApi.getSnacks();
      setSnacks(snacks);
    }
  }

  /**
   * Function to handle form submission of new item addition
   * pings API with item details LIKE {name, description, recipe, serve} and accepts TYPE LIKE "snacks"/"drinks"
   * updates data on site of the according type
   */
  async function addMenuItem(itemDetails, type) {
    await SnackOrBoozeApi.addNewItem(itemDetails, type);
    await updateData(type);
  };

  if (isLoading) {
    return <p>Loading!</p>;
  }

  return (
    <div className="App container-fluid">
      <BrowserRouter>
        <NavBar />
        <main>
          <Switch>
            <Route exact path="/">
              <Home totalDrinks={drinks.length} totalSnacks={snacks.length} />
            </Route>
            <Route exact path="/snacks">
              <FoodMenu snacks={snacks} />
            </Route>
            <Route exact path="/snacks/:id">
              <ItemDetails items={snacks} type="snacks" />
            </Route>
            <Route exact path="/drinks">
              <DrinkMenu drinks={drinks} />
            </Route>
            <Route exact path="/drinks/:id">
              <ItemDetails items={drinks} type="drinks" />
            </Route>
            <Route exact path="/add">
              <MenuItemForm addMenuItem={addMenuItem} />
            </Route>
            <Route>
              <NotFound />
            </Route>
          </Switch>
        </main>
      </BrowserRouter>
    </div>
  );
}

export default App;
