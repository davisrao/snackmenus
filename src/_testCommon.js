const drinks = [
  {
    "id": "martini",
    "name": "Martini",
    "description": "An ice-cold, refreshing classic.",
    "recipe": "Mix 3 parts vodka & 1 part dry vermouth.",
    "serve": "Serve very cold, straight up."
  },
  {
    "id": "negroni",
    "name": "Negroni",
    "description": "A nice drink for a late night conversation.",
    "recipe": "Mix equal parts of gin, Campari, and sweet vermouth.",
    "serve": "Serve cold, either on the rocks or straight up."
  }
]


const snacks = [
  {
    "type": "snacks",
    "name": "testname",
    "description": "testdesc",
    "recipe": "testrecipe",
    "serve": "testserve",
    "id": "testid"
  },
  {
    "type": "snacks",
    "name": "123",
    "description": "this is the test description for 123",
    "recipe": "123",
    "serve": "123",
    "id": "123"
  }
]



export { snacks, drinks };