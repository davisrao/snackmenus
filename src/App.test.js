import React from "react";
import { render, waitFor, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import App from "./App";
import axios from "axios";
import { drinks, snacks } from "./_testCommon.js"

const WELCOME_TEXT = "Welcome to your favorite dive cafe!";

jest.mock("axios")

describe("Cafe routes", function () {

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders without crashing", async function () {
        axios.get.mockResolvedValueOnce({ data: drinks });
        axios.get.mockResolvedValueOnce({ data: snacks });
        const { getByText } = render(
            <MemoryRouter>
                <App />
            </MemoryRouter>
        );
        expect(axios.get).toHaveBeenCalledWith(`http://localhost:5000/drinks`);
        expect(axios.get).toHaveBeenCalledWith(`http://localhost:5000/snacks`);
        await waitFor(() => getByText(WELCOME_TEXT));

    });

    it("renders the home snapshot", async function () {
        axios.get.mockResolvedValueOnce({ data: drinks });
        axios.get.mockResolvedValueOnce({ data: snacks });
        const { asFragment, getByText } = render(
            <MemoryRouter>
                <App />
            </MemoryRouter>
        );
        await waitFor(() => getByText(WELCOME_TEXT));
        expect(axios.get).toHaveBeenCalledTimes(2);
        getByText(/2 snack/)
        getByText(/2 drink/)
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders a loading message before the data is loaded", async function () {
        axios.get.mockResolvedValueOnce({ data: drinks });
        axios.get.mockResolvedValueOnce({ data: snacks });
        const { getByText } = render(
            <MemoryRouter>
                <App />
            </MemoryRouter>
        );
        getByText("Loading!");
        await waitFor(() => getByText(WELCOME_TEXT));
        expect(axios.get).toHaveBeenCalledTimes(2);
    });

    it("renders the snack menu and can go back", async function () {
        axios.get.mockResolvedValueOnce({ data: drinks });
        axios.get.mockResolvedValueOnce({ data: snacks });
        const { getByText } = render(
            <MemoryRouter >
                <App />
            </MemoryRouter>
        );
        await waitFor(() => getByText(WELCOME_TEXT));
        fireEvent.click(getByText("Snacks"));
        await waitFor(() => getByText("Snack Menu"));
        expect(axios.get).toHaveBeenCalledTimes(2);
        fireEvent.click(getByText("This is the cafe name"));
        await waitFor(() => getByText(WELCOME_TEXT));
    });

    it("renders the drink menu", async function () {
        axios.get.mockResolvedValueOnce({ data: drinks });
        axios.get.mockResolvedValueOnce({ data: snacks });
        const { getByText } = render(
            <MemoryRouter>
                <App />
            </MemoryRouter>
        );
        await waitFor(() => getByText(WELCOME_TEXT));
        fireEvent.click(getByText("Drinks"));
        await waitFor(() => getByText("Drink Menu"));
        expect(axios.get).toHaveBeenCalledTimes(2);
        fireEvent.click(getByText("This is the cafe name"));
        await waitFor(() => getByText(WELCOME_TEXT));
    });


    it("renders a drink from the menu and goes home", async function () {
        axios.get.mockResolvedValueOnce({ data: drinks });
        axios.get.mockResolvedValueOnce({ data: snacks });
        const { getByText } = render(
            <MemoryRouter>
                <App />
            </MemoryRouter>
        );
        await waitFor(() => getByText(WELCOME_TEXT));
        fireEvent.click(getByText("Drinks"));
        await waitFor(() => getByText("Drink Menu"));
        expect(axios.get).toHaveBeenCalledTimes(2);
        fireEvent.click(getByText("Martini"));
        await waitFor(() => getByText("An ice-cold, refreshing classic."));
        fireEvent.click(getByText("This is the cafe name"));
        await waitFor(() => getByText(WELCOME_TEXT));
    });


    it("renders a snack from the menu and goes home", async function () {
        axios.get.mockResolvedValueOnce({ data: drinks });
        axios.get.mockResolvedValueOnce({ data: snacks });
        const { getByText } = render(
            <MemoryRouter>
                <App />
            </MemoryRouter>
        );
        await waitFor(() => getByText(WELCOME_TEXT));
        fireEvent.click(getByText("Snacks"));
        await waitFor(() => getByText("Snack Menu"));
        expect(axios.get).toHaveBeenCalledTimes(2);
        fireEvent.click(getByText("123"));
        await waitFor(() => getByText("this is the test description for 123"));
        fireEvent.click(getByText("This is the cafe name"));
        await waitFor(() => getByText(WELCOME_TEXT));
    });

    it("renders a snack from the menu and goes home", async function () {
        axios.get.mockResolvedValueOnce({ data: drinks });
        axios.get.mockResolvedValueOnce({ data: snacks });
        const { getByText } = render(
            <MemoryRouter>
                <App />
            </MemoryRouter>
        );
        await waitFor(() => getByText(WELCOME_TEXT));
        fireEvent.click(getByText("Add to Menu!"));
        await waitFor(() => getByText("Serving Instructions:"));
        expect(axios.get).toHaveBeenCalledTimes(2);
    });
});
