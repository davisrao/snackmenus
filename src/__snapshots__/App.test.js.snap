// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cafe routes renders the home snapshot 1`] = `
<DocumentFragment>
  <div
    class="App container-fluid"
  >
    <div>
      <nav
        class="navbar navbar-expand-md"
      >
        <a
          aria-current="page"
          class="navbar-brand active"
          href="/"
        >
          This is the cafe name
        </a>
        <ul
          class="ml-auto navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              href="/snacks"
            >
              Snacks
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              href="/drinks"
            >
              Drinks
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              href="/add"
            >
              Add to Menu!
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <main>
      <section
        class="col-md-8"
      >
        <div
          class="card"
        >
          <div
            class="text-center card-body"
          >
            <div
              class="card-title"
            >
              <h3
                class="font-weight-bold"
              >
                Welcome to your favorite dive cafe!
              </h3>
              <h5>
                Home of 2 drink and 2 snack options.
              </h5>
            </div>
          </div>
        </div>
      </section>
    </main>
  </div>
</DocumentFragment>
`;
